src/persist/src/operators/stream.rs

```diff
@@ -497,32 +505,48 @@ where
 
                                     let future = primary_write.seal(*pending_future.cap.time());
 
-                                    // It's very important to add this to `pending_futures` and not
-                                    // `new_pending` to ensure that we poll from it at least once,
-                                    // so that it has a reference to the waker/activator.
-                                    pending_futures.push(PendingSealFuture::PrimarySeal(
+                                    pending_futures.push_back(PendingSealFuture::PrimarySeal(
                                         SealFuture {
                                             cap: pending_future.cap,
                                             future,
                                         },
                                     ));
                                 }
                                 std::task::Poll::Ready(Err(e)) => {
-                                    let mut session = data_output.session(&pending_future.cap);
                                     let error = format!(
                                         "Error sealing {} (condition) up to {}: {}",
                                         &operator_name,
                                         pending_future.cap.time(),
                                         e
                                     );
-                                    log::error!("{}", error);
+                                    log::trace!("{}", error);
 
-                                    // TODO: make error retractable? Probably not...
-                                    session.give(Err((error, *pending_future.cap.time(), 1)));
+                                    log::trace!(
+                                        "Adding seal (condition) to queue again: {}",
+                                        pending_future.cap.time()
+                                    );
+
+                                    // We push to the back, because we assume that this future will
```

Comment:
Pushing to the back is kind of tricky for me to reason about because it violates the whole "seal requests are worked off in order"

it seems like theres one of two things that could be happening here iiuc

1. a seal request for time t failed, and no other seal requests are pending. then we need to retry the seal so that the system can make progress, and we need to not advance timestamps past t yet
2. a seal request for time t failed and also seal requests for a time t' > t is already in flight / already completed. We don't need to retry the seal to time t per se, because the seal to t' covers that case, but we do need to retain a capability for time t, because the system can't advance past t yet because we haven't yet observed a successful seal to time >= t yet.

is that right?

Reply:
I think if so, there may be a simpler solution where we don't hold onto capabilities for each pending future, and instead hold onto one capability for the condition, and one for the primary, and explicitly downgrade them as we receive successful seal results? I think maybe that could simplify the logic here, let me see if I can noodle on that and come up with something